### PR TITLE
Add onCacheError option

### PR DESCRIPTION
**Pull request by slukes**
**Please check if the PR fulfills these requirements**

- [x]  Followed the [Contributing](https://github.com/node-cache-manager/node-cache-manager/blob/main/CONTRIBUTING.md) guidelines.

- [x] Tests for the changes have been added (for bug fixes/features) with 100% code coverage.
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This PR addresses:

https://github.com/node-cache-manager/node-cache-manager/issues/643
https://github.com/node-cache-manager/node-cache-manager/issues/588
https://github.com/node-cache-manager/node-cache-manager/issues/51

By re-adding an option that [existed](https://github.com/node-cache-manager/node-cache-manager/pull/4/files) in caching.js to 'ignore' cache errors but was [removed](https://github.com/node-cache-manager/node-cache-manager/pull/203) during the migration to typescript / promises.

This is useful to applications that are using a distant cache store, to enable to application to continue to work, even when the cache backend is down.

### CODE REVIEW

**Comment by jaredwray** [comment]
@slukes - I would rather not have another config on this and since it is cache should we just add in EventEmitter3 and emit an error when it goes to a catch?

**Comment by slukes** [comment]
> @slukes - I would rather not have another config on this and since it is cache should we just add in EventEmitter3 and emit an error when it goes to a catch?

Hello @jaredwray I'm not sure how this is going to work!  I've created a ticket on our team backlog to dig into this further in the next couple of weeks

**Comment by jaredwray** [comment]
@slukes - what I am saying is that instead of doing an `ignoreErrors` property in the cache manager what we can do is enable you to do this instead:

```
import { caching } from 'cache-manager';

const cache = await caching('memory', {
  max: 100,
  ttl: 10 * 1000 /*milliseconds*/,
});

cache.on('error', (error) = > {
     console.log(error);
});
```

This would use events to emit an error and by default we would make all methods catch errors and not throw. This gives the resiliency shown in `keyv` that many love. 

Would that work for you?

File: src/multi-caching.ts [hunk 1]
```diff
@@ -26,20 +30,28 @@ export function multiCaching<Caches extends Cache[]>(
 		data: T,
 		ttl?: Milliseconds | undefined,
 	) => {
-		await Promise.all(caches.map(async cache => cache.set(key, data, ttl)));
+		try {
+      await Promise.all(caches.map(async cache => cache.set(key, data, ttl)));
```
Comment (slukes) [inline src/multi-caching.ts:34]: @AdrienDoctolib I think we should talk about this - might be better to use allSettled

File: test/caching.test.ts [hunk 2]
```diff
@@ -319,6 +319,40 @@ describe('caching', () => {
 				});
 			}
 		});
+
+		describe('error handling on wrap', () => {
+			it('receives an error when store.get() fails', async () => {
```
Comment (slukes) [inline test/caching.test.ts:324]:
```suggestion
			it('emits and error event when store.get() fails', async () => {
```

File: test/caching.test.ts [hunk 3]
```diff
@@ -319,6 +319,40 @@ describe('caching', () => {
...
+				cache.store.get = vi.fn().mockRejectedValue(error);
+
+				let errorMessage;
+				cache.on('error', error => {
+					errorMessage = error;
+				});
+
+				await cache.wrap(key, function_);
+
+				expect(errorMessage).not.toBeUndefined();
```
Comment (slukes) [inline test/caching.test.ts:337]:
```suggestion
				expect(errorMessage).not.toBeUndefined();
				expect(function_).toHaveBeenCalled();
```

File: test/caching.test.ts [hunk 4]
```diff
@@ -319,6 +319,40 @@ describe('caching', () => {
...
+			});
+
+			it('receives an error when store.set() fails', async () => {
```
Comment (slukes) [inline test/caching.test.ts:340]:
```suggestion
			it('emits an error event when store.set() fails', async () => {
```

File: test/multi-caching.test.ts [hunk 5]
```diff
@@ -233,6 +248,30 @@ describe('multiCaching', () => {
 				await multiCache.set(key, value);
 				await expect(multiCache.get(key)).resolves.toBeUndefined();
 			});
+
+			it('should receive error event on get failure', async () => {
```
Comment (slukes) [inline test/multi-caching.test.ts:252]:
```suggestion
			it('emits an error event when store.get() fails', async () => {
```

**Comment by slukes** [comment]
@jaredwray thanks to the work of @AdrienDoctolib we now have something that is working how you suggested.  Please let us know your feedback. 

We have tested in our application and all is well 👍

**Comment by jaredwray** [comment]
@AdrienDoctolib and @slukes looks like we are getting an error on compile time:

Screenshot 2024-04-03 at 11 09 29 AM

**Comment by slukes** [comment]
> @AdrienDoctolib and @slukes looks like we are getting an error on compile time:

Sorry about that - fixed!

**Pull request merged**